hydro_cli/src/core/hydroflow_crate/ports.rs

```diff
@@ -156,12 +199,54 @@ impl HydroflowPortConfig {
 }
 
 impl HydroflowSource for HydroflowPortConfig {
+    fn source_path(&self) -> SourcePath {
+        SourcePath::Direct(
+            self.service
+                .upgrade()
+                .unwrap()
+                .try_read()
+                .unwrap()
+                .on
+                .clone(),
+        )
+    }
+
     fn send_to(&mut self, sink: &mut dyn HydroflowSink) {
         let from = self.service.upgrade().unwrap();
+
+        let forward_res =
+            sink.instantiate(&SourcePath::Direct(from.try_read().unwrap().on.clone()));
+        if let Ok(instantiated) = forward_res {
+            self.record_server_config(instantiated());
+        } else {
+            drop(forward_res);
+            let mut from_write = from.try_write().unwrap();
+            let instantiated = sink
+                .instantiate_reverse(
+                    &from_write.on,
+                    Arc::new(HydroflowPortConfig {
+                        service: Arc::downgrade(&from),
+                        port: self.port.clone(),
+                        merge: false,
+                    }),
+                    &|p| p,
+                )
+                .unwrap();
+
+            assert!(!from_write.port_to_bind.contains_key(&self.port));
+            from_write
+                .port_to_bind
+                .insert(self.port.clone(), instantiated(sink.as_any_mut()));
+        }
+    }
+
+    fn record_server_config(&mut self, config: ServerConfig) {
+        let from = self.service.upgrade().unwrap();
         let mut from_write = from.try_write().unwrap();
-        from_write
-            .add_connection(&from, self.port.clone(), sink)
-            .unwrap();
+
+        // TODO(shadaj): if already in this map, we want to broadcast
+        assert!(!from_write.port_to_server.contains_key(&self.port));
+        from_write.port_to_server.insert(self.port.clone(), config);
```

Comment:
todo?

Reply:
same TODO as elsewhere, need to address in a separate PR though...